pkg/kubectl/history.go

```diff
@@ -0,0 +1,139 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strconv"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_1"
+	"k8s.io/kubernetes/pkg/runtime"
+	deploymentutil "k8s.io/kubernetes/pkg/util/deployment"
+	"k8s.io/kubernetes/pkg/util/errors"
+)
+
+const (
+	ChangeCauseAnnotation = "kubernetes.io/change-cause"
+)
+
+// HistoryViewer provides an interface for resources that can be rolled back.
+type HistoryViewer interface {
+	History(namespace, name string) (HistoryInfo, error)
+}
+
+func HistoryViewerFor(kind unversioned.GroupKind, c clientset.Interface) (HistoryViewer, error) {
+	switch kind {
+	case extensions.Kind("Deployment"):
+		return &DeploymentHistoryViewer{c}, nil
+	}
+	return nil, fmt.Errorf("no history viewer has been implemented for %q", kind)
+}
+
+// HistoryInfo stores the mapping from revision to podTemplate;
+// note that change-cause annotation should be copied to podTemplate
+type HistoryInfo struct {
+	RevisionToTemplate map[int64]*api.PodTemplateSpec
+}
+
+type DeploymentHistoryViewer struct {
+	c clientset.Interface
+}
+
+// History returns a revision-to-RC map as the revision history of a deployment
+func (h *DeploymentHistoryViewer) History(namespace, name string) (HistoryInfo, error) {
+	historyInfo := HistoryInfo{
+		RevisionToTemplate: make(map[int64]*api.PodTemplateSpec),
+	}
+	deployment, err := h.c.Extensions().Deployments(namespace).Get(name)
+	if err != nil {
+		return historyInfo, fmt.Errorf("failed to retrieve deployment %s: %v", name, err)
+	}
+	_, allOldRCs, err := deploymentutil.GetOldRCs(*deployment, h.c)
+	if err != nil {
+		return historyInfo, fmt.Errorf("failed to retrieve old RCs from deployment %s: %v", name, err)
+	}
+	newRC, err := deploymentutil.GetNewRC(*deployment, h.c)
+	if err != nil {
+		return historyInfo, fmt.Errorf("failed to retrieve new RC from deployment %s: %v", name, err)
+	}
+	allRCs := append(allOldRCs, newRC)
+	for _, rc := range allRCs {
+		v, err := deploymentutil.Revision(rc)
+		if err != nil {
+			return historyInfo, fmt.Errorf("failed to retrieve revision out of RC %s from deployment %s: %v", rc.Name, name, err)
+		}
+		historyInfo.RevisionToTemplate[v] = rc.Spec.Template
+		changeCause, err := getChangeCause(rc)
+		if err != nil {
+			return historyInfo, fmt.Errorf("failed to retrieve change-cause out of RC %s from deployment %s: %v", rc.Name, name, err)
```

Comment:
Change cause is optional. We shouldn't bail on that. Could we just elide that column if it isn't present?